src/Simulation.cpp

```diff
@@ -988,6 +988,7 @@ void Simulation::simulate() {
         for (auto plugin : _plugins) {
             global_log -> debug() << "[BEFORE EVENT NEW TIMESTEP] Plugin: " << plugin->getPluginName() << endl;
             plugin->beforeEventNewTimestep(_moleculeContainer, _domainDecomposition, _simstep);
+			_domain->updateglobalNumMolecules(_moleculeContainer, _domainDecomposition);
         }
```

Comment:
Wasn't @pramathe's idea to call this not between the plugins but in `Domain::getglobalNumMolecules()`? That way we would only update it if we really need the value.

Reply:
Can you point me to the locations where this is only called by root? Is it feasible to either change the logic there so that the call is done for every rank? I assume that the particle numbers are also not correct if this is done without the allreduce.

Reply:
It is only called by root, e.g. in 

- [Domain.cpp:811](https://github.com/ls1mardyn/ls1-mardyn/blob/21809c642b32102d4215e734476a77d40a95f02e/src/Domain.cpp#L811)
- [Domain.cpp:570](https://github.com/ls1mardyn/ls1-mardyn/blob/21809c642b32102d4215e734476a77d40a95f02e/src/Domain.cpp#L570)
- [MmspdBinWriter.cpp:233](https://github.com/ls1mardyn/ls1-mardyn/blob/21809c642b32102d4215e734476a77d40a95f02e/src/io/MmspdBinWriter.cpp#L233)
- [ResultWriter.cpp:79](https://github.com/ls1mardyn/ls1-mardyn/blob/21809c642b32102d4215e734476a77d40a95f02e/src/io/ResultWriter.cpp#L71)

and probably a few others (like most of the readers and writers) ...
Plus the indirect calls eg. when calling `Domain::cv()`.

The `MPI_Allreduce` makes sure that the number of particles is updated in every rank, eventhough it may have been changed in only one rank.

Reply:
In all of those locations, there is directly an `if(rank != 0) ...`. it would be feasible to get the number of particles before that if, store it in a variable and then just use that (would also be better than calling `getglobalNumMolecules()` multiple times). Does this resolve all the deadlocks? What about the crash during reading the phase space? Is this related? Can you give more context?

Reply:
Ok so you already have your solutions to the problems ;) About the documentation I agree. Real propper testing is always hard and in ls1 we do way too little testing anyway so I don't see this as a reason for avoiding this.

Reply:
I just tried this solution and of course the update mechanism is called more often than just once per timestep, since the calculation of the pressure etc. require the total number of particles in the simulation and therefore use the `getglobalNumMolecules()` method. A flag does not make sense anymore as it would require the comparison of the old and new number of particles.
I will push the code and comment on some questionable lines.

Reply:
If you have the feeling that this is too slow the only option is using wrong (aka unupdated) numbers. No idea whether this makes sense but if it is just for pressure calculations it might be ok? In that case you could offer a parameter that indicates whether `getglobalNumMolecules()` should do communication or just use the last known value.

Reply:
I think, it is not _too_ slow, at least compared to other much more complicated calculations and communications in ls1 ;) Eventhough I will substitute the call `this->getglobalNumMolecules()` in `Domain::getAverageGlobalVirial()` and `Domain::getAverageGlobalUpot()` with the variable `_globalNumMolecules`.

Please also have a look at my comments below.